src/csharp/Microsoft.Spark/Utils/AssemblyLoader.cs

```diff
@@ -0,0 +1,126 @@
+
+        internal static Func<string, Assembly> LoadFromName { get; set; } = Assembly.Load;
+
+        private static readonly Dictionary<string, Assembly> s_assemblyCache =
+            new Dictionary<string, Assembly>();
+
+        private static readonly string[] s_searchPaths =
+            new[] { Directory.GetCurrentDirectory(), AppDomain.CurrentDomain.BaseDirectory };
+
+        private static readonly string[] s_extensions =
+            RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ?
+                    new[] { ".dll", ".exe", ".ni.dll", ".ni.exe" } :
+                    new[] { ".dll", ".ni.dll" };
+
+        private static readonly object s_cacheLock = new object();
+
+        /// <summary>
+        /// Return the cached assembly, otherwise attempt to load and cache the assembly
+        /// in the following order:
+        /// 1) Load the assembly from disk using assemblyFileName.
+        /// 2) Search for assemblyName in the current app domain.
+        /// </summary>
+        /// <param name="assemblyName">The full name of the assembly</param>
+        /// <param name="assemblyFileName">Name of the file that contains the assembly</param>
+        /// <returns>Cached or Loaded Assembly</returns>
+        internal static Assembly LoadAssembly(string assemblyName, string assemblyFileName)
+        {
+            lock (s_cacheLock)
+            {
+                if (s_assemblyCache.TryGetValue(assemblyName, out Assembly assembly))
+                {
+                    return assembly;
+                }
+
+                if (!TryLoadAssembly(assemblyFileName, ref assembly))
+                {
+                    assembly = LoadFromName(assemblyName);
+                }
+
+                s_assemblyCache[assemblyName] = assembly;
+                return assembly;
+            }
+        }
+
+        /// <summary>
+        /// Return the cached assembly, otherwise look in the following probing paths,
+        /// searching for the simple assembly name and s_extension combination.
+        /// 1) The working directory
+        /// 2) The directory of the application
+        /// </summary>
+        /// <param name="assemblyName">The fullname of the assembly to load</param>
+        /// <returns>The loaded assembly</returns>
+        /// <exception cref="FileNotFoundException">Thrown if the assembly is not
+        /// found.</exception>
+        internal static Assembly ResolveAssembly(string assemblyName)
+        {
+            lock (s_cacheLock)
+            {
+                if (s_assemblyCache.TryGetValue(assemblyName, out Assembly assembly))
+                {
+                    return assembly;
+                }
+
+                string simpleAsmName = new AssemblyName(assemblyName).Name;
+                foreach (string extension in s_extensions)
+                {
+                    var assemblyFileName = $"{simpleAsmName}{extension}";
```

Comment:
`string` I think.

Reply:
updated.  Although the type seems like it can be inferred based on the right hand value?

Reply:
yes, but then it allows doing `var i = 0`. Let me update the coding style guideline on this (I thought I updated it on `var`, but I guess I didn't.).